extensions/amp-list/0.1/amp-list.js

```diff
@@ -495,65 +492,104 @@ export class AmpList extends AMP.BaseElement {
+    if (this.element.hasAttribute('max-items')) {
+      items = this.truncateToMaxLen_(items);
+    }
+    return items;
+  }
+
+  /**
+   * Trigger a fetch-error event
+   * @param {!Error} error
+   */
+  triggerFetchErrorEvent_(error) {
+    const event = error
+      ? createCustomEvent(
+          this.win,
+          `${TAG}.error`,
+          dict({'response': error.response})
+        )
+      : null;
+    const actions = Services.actionServiceForDoc(this.element);
+    actions.trigger(this.element, 'fetch-error', event, ActionTrust.LOW);
+  }
   /**
    * Request list data from `src` and return a promise that resolves when
    * the list has been populated with rendered list items. If the viewer is
    * capable of rendering the templates, then the fetching of the list and
    * transformation of the template is handled by the viewer.
-   * @param {boolean=} opt_append
    * @param {boolean=} opt_refresh
    * @return {!Promise}
    * @private
    */
-  fetchList_(opt_append = false, opt_refresh = false) {
+  fetchList_(opt_refresh = false) {
     if (!this.element.getAttribute('src')) {
       return Promise.resolve();
     }
     let fetch;
     if (this.ssrTemplateHelper_.isSupported()) {
       fetch = this.ssrTemplate_(opt_refresh);
     } else {
-      const itemsExpr = this.element.getAttribute('items') || 'items';
       fetch = this.prepareAndSendFetch_(opt_refresh).then(data => {
-        let items = data;
-        if (itemsExpr != '.') {
-          items = getValueForExpr(/**@type {!JsonObject}*/ (data), itemsExpr);
-        }
-        userAssert(
-          typeof items !== 'undefined',
-          'Response must contain an array or object at "%s". %s',
-          itemsExpr,
-          this.element
-        );
-        if (this.element.hasAttribute('single-item') && !isArray(items)) {
-          items = [items];
-        }
-        items = user().assertArray(items);
-        if (this.element.hasAttribute('max-items')) {
-          items = this.truncateToMaxLen_(items);
-        }
+        const items = this.computeListItems_(data);
         if (this.loadMoreEnabled_) {
           this.updateLoadMoreSrc_(/** @type {!JsonObject} */ (data));
         }
-        return this.scheduleRender_(items, opt_append, /* opt_payload */ data);
+        return this.scheduleRender_(items, /*opt_append*/ false, data).then(
+          () => this.maybeSetLoadMore_()
+        );
       });
     }
 
     return fetch.catch(error => {
-      const event = error
-        ? createCustomEvent(
-            this.win,
-            `${TAG}.error`,
-            dict({'response': error.response})
-          )
-        : null;
-      const actions = Services.actionServiceForDoc(this.element);
-      actions.trigger(this.element, 'fetch-error', event, ActionTrust.LOW);
-
-      if (opt_append) {
-        throw error;
-      }
-      this.showFallbackOrThrow_(error);
+      this.triggerFetchErrorEvent_(error);
+      this.showFallback_();
+    });
```

Comment:
The two failing tests can be fixed by adding a 
```
throw error;
```
here below `this.showFallback();`. However, I'm unsure that this is the desired behavior--should we actually throw an error here and reject the layoutCallback promise? Or should we just silently show a fallback and change the test to reflect that behavior? 

```
            ● should fail to load b/c data array is absent
              Chrome 78.0.3904 (Mac OS X 10.14.6)
            AssertionError: expected promise to be rejected with an error matching /Response must contain an array/ but it was fulfilled with undefined

            ● should fail to load b/c data single-item object is absent
              Chrome 78.0.3904 (Mac OS X 10.14.6)
            AssertionError: expected promise to be rejected with an error matching /Response must contain an array or object/ but it was fulfilled with undefined
```

Reply:
We should not rethrow the error. Returning a rejected promise tells Resources to stop trying to interact with this element.

Reply:
> Returning a rejected promise tells Resources to stop trying to interact with this element.

Could you clarify what that means? 

Basically if the fetch fails here, yes, the layout failed because the fetch failed. But users can potentially retry the fetch through a variety of actions. So yes this is the end of the layoutCallback and yes it did fail, but we may try to layout this element again after a user interaction (e.g. bind src, refresh, etc.).

Reply:
Meaning `layoutCallback` won't be called again until `unlayoutCallback` is called. It gets stuck in an errored state. It can still receive actions, will still receive `viewport` events, etc. Just layouts will be errored.

Reply:
I'm still a little bit undecided on the behavior here on whether we should reject the layoutCallback promise given that we display a fallback. @choumx do you have an opinion on this? I've currently reverted my original decision here and brought the `throw error` back.

Reply:
Okay we reasoned about this: 
1. If the fetch fails on the first layout callback, we DON'T want it to relayout on viewport / size change. 
2. If the fetch fails on the first layout callback, and then we subsequently get a successful fetch based on a refresh or bind action, we DO want it to relayout on viewport / size change. 

So I guess ideally, we want this promise to reject layoutCallback, but we also want future successful fetches to take amp-list out of the errored layout state. Is there a way to do that?

Reply:
There's no current way to do that.

Reply:
Okay, in that case I'm going to leave this PR as is, without removing the `throw error` so that we do reject the promise if fetch fails.